### ISSUE TITLE
incorrect column length using  rows.ColumnTypes of database/sql

### ISSUE THREAD

**Issue by luxiangqing** [godror/godror#305]
**Describe the bug**
Using rows.ColumnTypes method to get varchar column length is 4 times the actual length,but using cx_Oracle with a similar method can get the correct result,the number type output results and cx_Oracle is not consistent, when using the default length to define the length of a column.

**To Reproduce**
 create table test (id number, name varchar2(100)) ;
```
func main() {
	query := "select id, name from test where rownum<=1"
	db, err := sqlx.Connect("godror", "xxxx")
	if err != nil {
		fmt.Println("connect database error: ", err)
		return
	}
	defer db.Close()
	cols, err := ColumnType(db, query)
	if err != nil {
		fmt.Println("query column type error: ", err)
		return
	}
	fmt.Println("%+v\n", cols)
}

func ColumnType(db *sqlx.DB, query string, args ...interface{}) (map[string]string, error) {
	types, err := db.QueryRowx(query, args...).ColumnTypes()
	if err != nil {
		return nil, err
	}
	m := make(map[string]string, len(types))
	for _, t := range types {
		colName := t.Name()
		colType := t.DatabaseTypeName()
		switch colType {
		case "VARCHAR2", "NCHAR", "NVARCHAR2", "CHAR":
			length, _ := t.Length()
			m[colName] = fmt.Sprintf("%s(%d)", colType, length)
		case "LONG", "LONG RAW", "CLOB", "NCLOB", "BLOB", "BFILE":
			m[colName] = colType
		case "DATE", "TIMESTAMP":
			m[colName] = colType
		case "NUMBER":
			precision, scale, _ := t.DecimalSize()
			m[colName] = fmt.Sprintf("%s(%d,%d)", colType, precision, scale)
		default:
			m[colName] = colType
		}
	}
	return m, nil
}
```
> godror output:  map[ID:NUMBER(0,-127) NAME:VARCHAR2(400)]
> cx_Oracle output: [('ID', <cx_Oracle.DbType DB_TYPE_NUMBER>, 127, None, 0, -127, 1), ('NAME', <cx_Oracle.DbType DB_TYPE_VARCHAR>, 100, 400, None, None, 1)]

**Expected behavior**

**Error output**
Copy the error output (panic, stack trace),
or the output of running with `DPI_DEBUG_LEVEL=92`.

**Your oracle client version**
19.13.0.0

**Your godror version**
v0.37.0

**Your go version**
1.20.7

**Comment by tgulacsi** [comment]
What is sqlx?

For table column types, use the user_tab_cols view.

ColumnTypes' goal is to show how much space is needed for that column.

So it returns the length in bytes, which can be 4 times the length of utf-8 runes ("characters").

GT
________________________________
From: luxiangqing ***@***.***>
Sent: Wednesday, August 16, 2023 6:10:06 AM
To: godror/godror ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [godror/godror] incorrect column length using rows.ColumnTypes of database/sql (Issue #305)


Describe the bug
Using rows.ColumnTypes method to get varchar column length is 4 times the actual length,but using cx_Oracle with a similar method can get the correct result,the number type output results and cx_Oracle is not consistent, when using the default length to define the length of a column.

To Reproduce
create table test (id number, name varchar2(100)) ;

func main() {
        query := "select id, name from test where rownum<=1"
        db, err := sqlx.Connect("godror", "xxxx")
        if err != nil {
                fmt.Println("connect database error: ", err)
                return
        }
        defer db.Close()
        cols, err := ColumnType(db, query)
        if err != nil {
                fmt.Println("query column type error: ", err)
                return
        }
        fmt.Println("%+v\n", cols)
}

func ColumnType(db *sqlx.DB, query string, args ...interface{}) (map[string]string, error) {
        types, err := db.QueryRowx(query, args...).ColumnTypes()
        if err != nil {
                return nil, err
        }
        m := make(map[string]string, len(types))
        for _, t := range types {
                colName := t.Name()
                colType := t.DatabaseTypeName()
                switch colType {
                case "VARCHAR2", "NCHAR", "NVARCHAR2", "CHAR":
                        length, _ := t.Length()
                        m[colName] = fmt.Sprintf("%s(%d)", colType, length)
                case "LONG", "LONG RAW", "CLOB", "NCLOB", "BLOB", "BFILE":
                        m[colName] = colType
                case "DATE", "TIMESTAMP":
                        m[colName] = colType
                case "NUMBER":
                        precision, scale, _ := t.DecimalSize()
                        m[colName] = fmt.Sprintf("%s(%d,%d)", colType, precision, scale)
                default:
                        m[colName] = colType
                }
        }
        return m, nil
}


godror output: map[ID:NUMBER(0,-127) NAME:VARCHAR2(400)]
cx_Oracle output: [('ID', <cx_Oracle.DbType DB_TYPE_NUMBER>, 127, None, 0, -127, 1), ('NAME', <cx_Oracle.DbType DB_TYPE_VARCHAR>, 100, 400, None, None, 1)]

Expected behavior
A clear and concise description of what you expected to happen.

Error output
Copy the error output (panic, stack trace),
or the output of running with DPI_DEBUG_LEVEL=92.

Your oracle client version
19.13.0.0

Your godror version
v0.37.0

Your go version
1.20.7

**Comment by tgulacsi** [comment]
A simple "UPPER(col1)" will erase the direct link of the column and the query's column, so you can't count on it either.
So for a query, the column will have a type and some kind of length. This length may mean several things:
a) size in bytes
b) size in runes (characters)
c) normalized size in runes.

From these, a) has been chosen as the easiest, simplest and most useful (the storage size for that column).

To elaborate a little bit: on a query, all you can know is that column is a VARCHAR2, 100 characters wide
(dpiDataTypeInfo.sizeInChars is 100 at https://odpi-c.readthedocs.io/en/latest/structs/dpiDataTypeInfo.html#dpidatatypeinfo).
But that's characters.
Ok, you know the DB storage size, too (dpiDataTypeInfo.dbSizeInBytes), but that's not what you get on the wire.
You get dpiDataTypeInfo.clientSizeInBytes, UTF-8 (that's what we use on the client size, as that's Go's native encoding).

So you'll need a 400 byte length storage on the client side, but the column may be "VARCHAR2(100 CHAR)" or "VARCHAR2(400 BYTE)" on the database, with any encoding (we know, but shall we calculate something on it?).

TL;DR; if you want to know the table's column type, use "the user_tab_cols" view.
